extensions/storage-image-processing-api/functions/src/index.ts

```diff
@@ -0,0 +1,164 @@
+import helmet from 'helmet';
+import * as functions from 'firebase-functions';
+import { StructError } from 'superstruct';
+import {
+  expressAsyncHandlerWrapper,
+  fileMetadataBufferKeys,
+  hasOwnProperty,
+  omitKeys,
+} from './utils';
+import { ValidatedOperation } from './types';
+import { extensionConfiguration } from './config';
+
+const app = express();
+
+app.use(helmet());
+app.use(
+  cors({
+    origin: extensionConfiguration.corsAllowList,
+    methods: ['GET', 'POST'],
+  }),
+);
+
+app.use(
+  '/process/**',
+  expressAsyncHandlerWrapper(async (req, res, next) => {
+    const operationsString = req.baseUrl.replace('/process/', '');
+    const queryParams = req.query;
+
+    // 404 if no operations
+    if (!operationsString || !operationsString.length) {
+      return next();
+    }
+
+    const validatedOperations: ValidatedOperation[] =
+      asValidatedOperations(operationsString);
+
+    // Ensure input is the first operation.
+    if (validatedOperations[0].operation != 'input') {
+      throw new AssertionError({
+        message: `An input operation must be the first operation.`,
+      });
+    }
+
+    // Ensure output is the last operation.
+    if (
+      validatedOperations[validatedOperations.length - 1].operation != 'output'
+    ) {
+      throw new AssertionError({
+        message: `An output operation must be the last operation.`,
+      });
+    }
+
+    // Apply operations.
+    let instance = null;
+    for (let i = 0; i < validatedOperations.length; i++) {
+      const validatedOperation = validatedOperations[i];
+      instance = await applyValidatedOperation(instance, validatedOperation);
+    }
+
+    if (hasOwnProperty(queryParams, 'debug')) {
+      const [metadataError, fileMetadata] = await a2a(instance.metadata());
+      if (metadataError) {
+        return next(metadataError);
+      }
+      return res.json({
+        operations: validatedOperations,
+        metadata: omitKeys(fileMetadata, fileMetadataBufferKeys),
+      });
+    }
+
+    const [outputError, output] = await a2a(
+      instance.toBuffer({ resolveWithObject: true }),
+    );
+    if (outputError) {
+      return next(outputError);
+    }
+
+    const { data, info } = output;
+
+    functions.logger.debug(`Processed a new request.`, validatedOperations);
+
+    const headers = {
+      'Content-Type': `image/${info.format}`,
+      'Content-Length': data.length,
+      'Content-Disposition': `inline; filename=image.${info.format}`,
+    };
+
+    Object.entries(info).forEach(entry => {
+      headers[`ext-output-info-${entry[0].toLowerCase()}`] = `${entry[1]}`;
+    });
+
+    // TODO attach headers using final fileMetadata
+
+    res.writeHead(200, headers);
+    res.end(data);
+  }),
+);
+
+// Express requires the function to have 4 arguments for a handler
+// to be treated as an error handler.
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+app.use(function (error, req, res, next) {
```

Comment:
nitpick: May help to name this function so its purpose is immediately clear

```suggestion
app.use(function handleError(error, req, res, next) {
```

Reply:
Great suggestion, thanks!